File: backend/workload/generator.py
"""
Workload & Traffic Generator
Generates realistic Zipfian/Pareto distributed request streams with dynamic scenario states.
"""

import math
import random
import time
from typing import Dict, Any, List, Optional
from backend.core.types import WorkloadType, TrafficScenario, SystemConfig
from backend.workload.catalog import ItemCatalog


class WorkloadGenerator:
    def __init__(self, catalog: Optional[ItemCatalog] = None):
        self.catalog = catalog or ItemCatalog()
        self.workload_type = WorkloadType.READ_HEAVY_API
        self.scenario = TrafficScenario.STEADY
        self.base_rps = SystemConfig.baseRps
        self.speed_multiplier = 1.0

        self.sim_time_seconds = 0.0
        self.spike_target_item_id = "prod_048"
        self.shift_phase = 0
        self.pollution_rate = 0.0

        self.zipf_weights: Dict[str, float] = {}
        self.recompute_weights()

    def set_workload_type(self, w_type: WorkloadType):
        if self.workload_type != w_type:
            self.workload_type = w_type
            self.spike_target_item_id = "rec_model_012" if w_type == WorkloadType.COMPUTE_HEAVY_REC else "prod_048"
            self.recompute_weights()

    def set_scenario(self, scenario: TrafficScenario, options: Optional[Dict[str, Any]] = None):
        opts = options or {}
        self.scenario = scenario
        if "targetItemId" in opts:
            self.spike_target_item_id = opts["targetItemId"]
        
        if scenario == TrafficScenario.CACHE_POLLUTION:
            self.pollution_rate = opts.get("pollutionRate", 0.85)
        else:
            self.pollution_rate = 0.0

        self.recompute_weights()

    def set_base_rps(self, rps: int):
        self.base_rps = max(10, min(5000, int(rps)))

    def set_speed_multiplier(self, multiplier: float):
        self.speed_multiplier = max(0.1, float(multiplier))

    def recompute_weights(self):
        items = self.catalog.get_catalog(self.workload_type)
        n = len(items)
        s = 1.25 if self.workload_type == WorkloadType.READ_HEAVY_API else 0.85

        self.zipf_weights.clear()
        total_weight = 0.0

        for rank in range(1, n + 1):
            item = items[rank - 1]
            weight = 1.0 / math.pow(rank, s)

            if self.scenario == TrafficScenario.POPULARITY_SPIKE and item["id"] == self.spike_target_item_id:
                weight *= 45.0  # 45x surge on spike target
            elif self.scenario == TrafficScenario.GRADUAL_SHIFT:
                shifted_rank = ((rank - 1 + self.shift_phase) % n) + 1
                weight = 1.0 / math.pow(shifted_rank, s)

            self.zipf_weights[item["id"]] = weight
            total_weight += weight

        # Normalize probabilities
        for item_id in self.zipf_weights:
            self.zipf_weights[item_id] /= max(1e-9, total_weight)

    def tick(self, delta_ms: int) -> Dict[str, Any]:
        self.sim_time_seconds += (delta_ms / 1000.0) * self.speed_multiplier

        if self.scenario == TrafficScenario.GRADUAL_SHIFT:
            items = self.catalog.get_catalog(self.workload_type)
            self.shift_phase = int(self.sim_time_seconds / 15.0) % len(items)
            self.recompute_weights()

        effective_rps = float(self.base_rps)
        if self.scenario == TrafficScenario.TRAFFIC_BURST:
            effective_rps *= 3.8
        elif self.scenario == TrafficScenario.COLD_START:
            effective_rps *= 1.2

        request_count = max(1, int(round((effective_rps * (delta_ms / 1000.0)) * self.speed_multiplier)))
        requests = []

        for _ in range(request_count):
            if self.scenario == TrafficScenario.CACHE_POLLUTION and random.random() < self.pollution_rate:
                unique_key_id = f"crawler_scan_{int(time.time()*1000)}_{random.randint(1000, 999999)}"
                requests.append({
                    "id": unique_key_id,
                    "name": f"Uncached Ephemeral Key ({unique_key_id[-6:]})",
                    "category": "EPHEMERAL_BOT",
                    "type": "UncachedProbe",
                    "sizeBytes": 8 * 1024 + random.randint(0, 64 * 1024),
                    "baseDbLatencyMs": 35.0 + random.randint(0, 50),
                    "recomputeCostUnits": 1.0,
                    "updateVolatility": 1.0,
                    "isPollutionKey": True,
                    "timestamp": self.sim_time_seconds
                })
                continue

            item = self._sample_item()
            if item:
                req_copy = dict(item)
                req_copy["timestamp"] = self.sim_time_seconds
                requests.append(req_copy)

        return {
            "simTimeSeconds": self.sim_time_seconds,
            "workloadType": self.workload_type.value,
            "scenario": self.scenario.value,
            "effectiveRps": round(effective_rps, 1),
            "requestCount": len(requests),
            "requests": requests
        }

    def _sample_item(self) -> Optional[Dict[str, Any]]:
        r = random.random()
        cumulative = 0.0
        items = self.catalog.get_catalog(self.workload_type)

        for item in items:
            prob = self.zipf_weights.get(item["id"], 0.0)
            cumulative += prob
            if r <= cumulative:
                return item

        return items[-1] if items else None

File: src/workload/generator.js
/**
 * Workload & Traffic Generator
 * Generates realistic Zipfian/Pareto distributed request streams with dynamic scenario states
 */

import { WORKLOAD_TYPES, TRAFFIC_SCENARIOS, DEFAULT_CONFIG } from '../core/types.js';
import { ItemCatalog } from './item-catalog.js';

export class WorkloadGenerator {
  constructor(itemCatalog = new ItemCatalog()) {
    this.catalog = itemCatalog;
    this.workloadType = WORKLOAD_TYPES.READ_HEAVY_API;
    this.scenario = TRAFFIC_SCENARIOS.STEADY;
    this.baseRps = DEFAULT_CONFIG.baseRps;
    this.speedMultiplier = 1;

    // Internal simulation time tracking
    this.simTimeSeconds = 0;
    this.spikeTargetItemId = 'prod_048';
    this.shiftPhase = 0;
    this.pollutionRate = 0; // 0 (none) to 1.0 (heavy flood)

    // Precomputed Zipfian weights
    this.zipfWeights = new Map();
    this.recomputeWeights();
  }

  setWorkloadType(type) {
    if (this.workloadType !== type) {
      this.workloadType = type;
      this.spikeTargetItemId = (type === WORKLOAD_TYPES.COMPUTE_HEAVY_REC) ? 'rec_model_012' : 'prod_048';
      this.recomputeWeights();
    }
  }

  setScenario(scenario, options = {}) {
    this.scenario = scenario;
    if (options.targetItemId) this.spikeTargetItemId = options.targetItemId;
    if (scenario === TRAFFIC_SCENARIOS.CACHE_POLLUTION) {
      this.pollutionRate = options.pollutionRate || 0.85;
    } else {
      this.pollutionRate = 0;
    }
    this.recomputeWeights();
  }

  setBaseRps(rps) {
    this.baseRps = Math.max(10, Math.min(5000, Number(rps)));
  }

  setSpeedMultiplier(multiplier) {
    this.speedMultiplier = Number(multiplier) || 1;
  }

  recomputeWeights() {
    const items = this.catalog.getCatalog(this.workloadType);
    const n = items.length;
    let s = (this.workloadType === WORKLOAD_TYPES.READ_HEAVY_API) ? 1.25 : 0.85; // Skew factor
    
    this.zipfWeights.clear();
    let sum = 0;

    for (let rank = 1; rank <= n; rank++) {
      const item = items[rank - 1];
      let weight = 1 / Math.pow(rank, s);

      // Scenario adjustments
      if (this.scenario === TRAFFIC_SCENARIOS.POPULARITY_SPIKE && item.id === this.spikeTargetItemId) {
        weight *= 45.0; // 45x spike
      } else if (this.scenario === TRAFFIC_SCENARIOS.GRADUAL_SHIFT) {
        // Shift hot window based on shiftPhase (0 -> n)
        const shiftedRank = ((rank - 1 + this.shiftPhase) % n) + 1;
        weight = 1 / Math.pow(shiftedRank, s);
      }

      this.zipfWeights.set(item.id, weight);
      sum += weight;
    }

    // Normalize probabilities
    for (const [id, weight] of this.zipfWeights.entries()) {
      this.zipfWeights.set(id, weight / sum);
    }
  }

  tick(deltaMs) {
    this.simTimeSeconds += (deltaMs / 1000) * this.speedMultiplier;

    // Handle Gradual Shift progress
    if (this.scenario === TRAFFIC_SCENARIOS.GRADUAL_SHIFT) {
      const items = this.catalog.getCatalog(this.workloadType);
      this.shiftPhase = Math.floor(this.simTimeSeconds / 15) % items.length;
      this.recomputeWeights();
    }

    // Effective RPS calculation based on scenario
    let effectiveRps = this.baseRps;
    if (this.scenario === TRAFFIC_SCENARIOS.TRAFFIC_BURST) {
      // 4x Burst
      effectiveRps *= 3.8;
    } else if (this.scenario === TRAFFIC_SCENARIOS.COLD_START) {
      effectiveRps *= 1.2;
    }

    const requestCount = Math.max(1, Math.round((effectiveRps * (deltaMs / 1000)) * this.speedMultiplier));
    const requests = [];

    // Generate specific item requests based on distribution
    for (let i = 0; i < requestCount; i++) {
      // Check if pollution key should be injected
      if (this.scenario === TRAFFIC_SCENARIOS.CACHE_POLLUTION && Math.random() < this.pollutionRate) {
        const uniqueKeyId = `crawler_scan_${Date.now()}_${Math.floor(Math.random() * 1000000)}`;
        requests.push({
          id: uniqueKeyId,
          name: `Uncached Ephemeral Key (${uniqueKeyId.slice(-6)})`,
          category: 'EPHEMERAL_BOT',
          type: 'UncachedProbe',
          sizeBytes: 8 * 1024 + Math.floor(Math.random() * 64 * 1024),
          baseDbLatencyMs: 35 + Math.floor(Math.random() * 50),
          recomputeCostUnits: 1.0,
          updateVolatility: 1.0,
          isPollutionKey: true,
          timestamp: this.simTimeSeconds
        });
        continue;
      }

      // Sample from Zipfian catalog
      const item = this.sampleItem();
      if (item) {
        requests.push({
          ...item,
          timestamp: this.simTimeSeconds
        });
      }
    }

    return {
      simTimeSeconds: this.simTimeSeconds,
      workloadType: this.workloadType,
      scenario: this.scenario,
      effectiveRps,
      requestCount,
      requests
    };
  }

  sampleItem() {
    const r = Math.random();
    let cumulative = 0;
    const items = this.catalog.getCatalog(this.workloadType);

    for (const item of items) {
      const prob = this.zipfWeights.get(item.id) || 0;
      cumulative += prob;
      if (r <= cumulative) {
        return item;
      }
    }
    return items[items.length - 1];
  }
}
